integration-tests/cucumber-tests-protocol-iec60870/src/test/java/org/opensmartgridplatform/cucumber/protocol/iec60870/domain/AbstractIec60870DeviceCreator.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2021 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.cucumber.protocol.iec60870.domain;
+
+import java.util.Map;
+
+import org.opensmartgridplatform.adapter.protocol.iec60870.domain.entities.Iec60870Device;
+import org.opensmartgridplatform.adapter.protocol.iec60870.domain.valueobjects.DeviceType;
+import org.opensmartgridplatform.cucumber.core.ReadSettingsHelper;
+import org.opensmartgridplatform.cucumber.platform.helpers.ProtocolDeviceCreator;
+
+public abstract class AbstractIec60870DeviceCreator implements ProtocolDeviceCreator<Iec60870Device> {
+
+    private static final String KEY_DEVICE_IDENTIFICATION = "DeviceIdentification";
+    private static final String KEY_GATEWAY_DEVICE_IDENTIFICATION = "GatewayDeviceIdentification";
+    private static final String KEY_COMMON_ADDRESS = "CommonAddress";
+    private static final String KEY_PORT = "Port";
+    private static final String KEY_INFORMATION_OBJECT_ADDRESS = "InformationObjectAddress";
+
+    private static final String DEFAULT_DEVICE_IDENTIFICATION = "IEC60870-DVC-1";
+    private static final String DEFAULT_GATEWAY_DEVICE_IDENTIFICATION = "IEC60870-GW-1";
+    private static final int DEFAULT_COMMON_ADDRESS = 1;
+    private static final int DEFAULT_PORT = 62404;
+    private static final int DEFAULT_INFORMATION_OBJECT_ADDRESS = 1;
+
+    @Override
+    public abstract Iec60870Device apply(Map<String, String> settings);
+
+    protected abstract DeviceType deviceType();
+
+    protected String deviceIdentification(final Map<String, String> settings) {
+        return ReadSettingsHelper.getString(settings, KEY_DEVICE_IDENTIFICATION, this.defaultDeviceIdentification());
+    }
+
+    protected String gatewayDeviceIdentification(final Map<String, String> settings) {
+        return ReadSettingsHelper.getString(settings, KEY_GATEWAY_DEVICE_IDENTIFICATION,
+                this.defaultGatewayDeviceIdentification());
+    }
+
+    protected Integer port(final Map<String, String> settings) {
+        return ReadSettingsHelper.getInteger(settings, KEY_PORT, this.defaultPort());
+    }
+
+    protected Integer commonAddress(final Map<String, String> settings) {
+        return ReadSettingsHelper.getInteger(settings, KEY_COMMON_ADDRESS, this.defaultCommonAddress());
+    }
+
+    protected Integer informationObjectAddress(final Map<String, String> settings) {
+        return ReadSettingsHelper.getInteger(settings, KEY_INFORMATION_OBJECT_ADDRESS,
+                this.defaultInformationObjectAddress());
+    }
+
+    protected String defaultDeviceIdentification() {
```

Comment:
The main difference between deviceType and for example defaultDeviceIdentification is that DeviceType is fixed for each subclass, the defaultDeviceIdentification provides a default that may be overridden in a subclass, as well as by the parameters in the step definition. 
Perhaps these options to override defaults could have been achieved in another way as well...

Reply:
I wonder if we disagree or if I did not express my comment clearly enough.

I think this would make sense to provide a default in the abstract base class (based on my expectations):
protected String deviceIdentification() {
    return DEFAULT_DEVICE_IDENTIFICATION;
}

Adding "default" in the name of the method seems odd to me. I have a hard time understanding why abstract methods would be named "deviceType", but if a default were available, the method name would change to "defaultDeviceType".

What I would expect is that subclasses would either refer to the constants directly (where the constants having a name that starts with "DEFAULT_" seems a logical choice), or would inherit the default value from the implementation in the base class. In this latter case I find it unexpected that the method is named "defaultDeviceIdentification", since the way I looked at it the interface could just be "deviceIdentification".

If I missed something and there is another reason the word default is used in the method with device identification and not with device type, that could be OK. The only point then would be if this could be made more clear so I wouldn't get confused so easily.

Reply:
I used the prefix "default" to indicate these are default values that can be overridden by the settings parameter. In contrast to the deviceType which is fixed for a subclass.

Thus for most values: Default value of base class -> Default value of sub class -> Value from settings
For DeviceType only: No value in base class -> Fixed value in sub class

But I have no problem dropping the "default" prefix if it makes things clearer.